### ISSUE TITLE
Index Checker: failed to recognize local variables, issued wrong required type.

### ISSUE THREAD

**Issue by NhatDinh** [typetools/checker-framework#1913]
Index checker should prompt required type for local and parameter variables as accurate as field variables. Consider:
```java
public class SimpleMethod1 {
    int[] fields_arr = new int[10];
    void simpleMethod( int i, int[] para_arr) {
        int[] local_arr = new int[10];
        this.fields_arr[i] = i;
        para_arr[i] = i;
        local_arr[i] = i;
    }
    public static void main(String args[]) {
    }
}
```
Warnings received using command line:
```java
SimpleMethod1.java:9: error: [array.access.unsafe.low] Potentially unsafe array access: the index could be negative.
        this.fields_arr[i] = i;
                        ^
  found   : @LowerBoundUnknown int
  required: an integer >= 0 (@NonNegative or @Positive)
SimpleMethod1.java:9: error: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
        this.fields_arr[i] = i;
                        ^
  found   : @UpperBoundUnknown int
  required: @IndexFor("this.fields_arr") or @LTLengthOf("this.fields_arr") -- an integer less than this.fields_arr's length
SimpleMethod1.java:10: error: [array.access.unsafe.low] Potentially unsafe array access: the index could be negative.
        para_arr[i] = i;
                 ^
  found   : @LowerBoundUnknown int
  required: an integer >= 0 (@NonNegative or @Positive)
SimpleMethod1.java:10: error: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
        para_arr[i] = i;
                 ^
  found   : @UpperBoundUnknown int
  required: @IndexFor("para_arr") or @LTLengthOf("para_arr") -- an integer less than para_arr's length
SimpleMethod1.java:11: error: [array.access.unsafe.low] Potentially unsafe array access: the index could be negative.
        local_arr[i] = i;
                  ^
  found   : @LowerBoundUnknown int
  required: an integer >= 0 (@NonNegative or @Positive)
SimpleMethod1.java:11: error: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
        local_arr[i] = i;
                  ^
  found   : @UpperBoundUnknown int
  required: @IndexFor("local_arr") or @LTLengthOf("local_arr") -- an integer less than local_arr's length
```
In this case, Index Checker prompted the wrong required type, adding annotations accordingly will lead to identifiers not found.
+ For fields_arr, @IndexFor("this.fileds_arr") is correct
+ For para_arr, correct required type should be: @IndexFor("#2). Java doc should have more examples of annotating index of parameters arrays.
+ For local_arr, I could not find the right required type for this in the manual and Javadoc.

This is the reason why I believe Java doc should include more examples( code snippets) showcasing how value in @IndexFor() should be annotated in different scenarios.

**Comment by kelloggm** [comment]
Thanks for reporting your confusion with these error reports. However, I don't believe that the errors are incorrect, because the types they prompt for are the required types _at the program point at which they are issued_. In other words, the error doesn't report the annotation you need to write on `i`, because in general the correct annotation can't be decided automatically (that's why we ask programmers to write them). Instead, it reports the type that `i` needs to have in order to verify the array access at that point in the program. So, for the access to `para_arr`, `i`'s type needs to be `@IndexFor("para_arr")`, because `para_arr` is in scope there (the `#2` syntax is required because Java bytecode can't contain parameter names, which is a limitation of Java - if that weren't so, we could use `@IndexFor("para_arr")` instead of `@IndexFor("#2")` in the signature of the function, too).

All that said, I agree that the Index Checker's error messages could be better, and that the documentation could be more clear. I'm not sure exactly how to address this in the manual, though, since this is a fundamental property of how all Checker Framework type systems work.

**Comment by mernst** [comment]
The need for the "#2" syntax is explained at https://checkerframework.org/manual/#java-expressions-as-arguments .
Concrete suggestions for improving the Index Checker's error messages would be quite welcome.

**Comment by NhatDinh** [comment]
@kelloggm @mernst Thank you for explaining, now I understand that the prompted required types don't have to report the correct annotation. 
However, I think programmers should be informed about the needed type in error messages, so they only need to look for the explanation( i.e "Java bytecode can't contain parameter names, which is a limitation of Java"), not the correct annotation.
As for  the "#2" syntax, my suggestions( drawn from GSoC "Improving error messages") is to include one line as "suggested" types: 
```java
 found   : @UpperBoundUnknown int
 required: @IndexFor("local_arr") or @LTLengthOf("local_arr") -- an integer less than local_arr's length
 suggested: @IndexFor("#2") or @LTLengthOf("#2") -- #2 as local_arr
```
Can you also point me to and explain the correct way to annotate @IndexFor("local_arr") in this case?
Let me know your opinions.
